instrumentation/log4j/log4j-appender-2.17/library/src/main/java/io/opentelemetry/instrumentation/log4j/appender/v2_17/internal/LogEventMapper.java

```diff
@@ -169,51 +185,200 @@ void captureMessage(LogRecordBuilder builder, AttributesBuilder attributes, Mess
-      mapMessage
-          .getData()
-          .forEach(
-              (key, value) -> {
-                if (value != null
-                    && (!checkSpecialMapMessageAttribute
-                        || !key.equals(SPECIAL_MAP_MESSAGE_ATTRIBUTE))) {
-                  attributes.put(getMapMessageAttributeKey(key), value.toString());
-                }
-              });
+      consumeAttributes(
+          mapMessage.getData()::forEach,
+          attributes,
+          checkSpecialMapMessageAttribute,
+          LogEventMapper::getMapMessageAttributeKey);
     }
   }
 
-  // visible for testing
-  void captureContextDataAttributes(AttributesBuilder attributes, T contextData) {
+  private static void consumeAttributes(
+      // Consumes an action on an entry, like map::forEach
+      Consumer<BiConsumer<String, Object>> entryActionConsumer,
+      ExtendedAttributesBuilder attributes,
+      boolean checkSpecialMapMessageAttribute,
+      BiFunction<String, ExtendedAttributeType, ExtendedAttributeKey<?>> keyProvider) {
+    entryActionConsumer.accept(
+        (key, value) -> {
+          if (value != null
+              && (!checkSpecialMapMessageAttribute || !key.equals(SPECIAL_MAP_MESSAGE_ATTRIBUTE))) {
+            consumeEntry(key, value, attributes, keyProvider);
+          }
+        });
+  }
+
+  @SuppressWarnings({"unchecked"})
+  private static void consumeEntry(
+      String key,
+      Object value,
+      ExtendedAttributesBuilder attributes,
+      BiFunction<String, ExtendedAttributeType, ExtendedAttributeKey<?>> keyProvider) {
+    if (value instanceof String) {
+      attributes.put(
+          (ExtendedAttributeKey<String>) keyProvider.apply(key, ExtendedAttributeType.STRING),
+          (String) value);
+    } else if (value instanceof Boolean) {
+      attributes.put(
+          (ExtendedAttributeKey<Boolean>) keyProvider.apply(key, ExtendedAttributeType.BOOLEAN),
+          (Boolean) value);
+    } else if ((value instanceof Long)) {
+      attributes.put(
+          (ExtendedAttributeKey<Long>) keyProvider.apply(key, ExtendedAttributeType.LONG),
+          (Long) value);
+    } else if ((value instanceof Integer)) {
+      attributes.put(
+          (ExtendedAttributeKey<Long>) keyProvider.apply(key, ExtendedAttributeType.LONG),
+          ((Integer) value).longValue());
+    } else if (value instanceof Double) {
+      attributes.put(
+          (ExtendedAttributeKey<Double>) keyProvider.apply(key, ExtendedAttributeType.DOUBLE),
+          (Double) value);
+    } else if (value instanceof Float) {
+      attributes.put(
+          (ExtendedAttributeKey<Double>) keyProvider.apply(key, ExtendedAttributeType.DOUBLE),
+          ((Float) value).doubleValue());
+    } else if (value instanceof List) {
+      List<?> list = (List<?>) value;
+      if (list.isEmpty()) {
+        return;
+      }
 
+      Object first = list.get(0);
```

Comment:
I suspect something horrible will happen when list contains elements of mixed types.

Reply:
Here I made it fallback to a list of strings: 4f45fc4dc6879b46a6992f30ef837e3131ca9889